// raf-tasq/lib/consumer/consumer.js
const EventEmitter = require('events');
const Queue = require('bull');
const validate = require('djsv');
const schema = require('./schema');
const helper = require('../helpers/helper');

class Consumer extends EventEmitter {

    constructor(options) {
        super();
        this._setting = Object.assign({}, options.setting);
        const res = validate(schema.properties.setting, this._setting);
        if (!res.valid) {
            throw new Error(res.errors[0].stack);
        }
        this._queue = new Queue(this._setting.queueName, this._setting);
    }

    register(options) {
        options = options || {};
        const res = validate(schema, options);
        if (!res.valid) {
            throw new Error(res.errors[0].stack);
        }
        this._queue.process(options.job.type, 1, (job, done) => {
            const res = {
                id: job.id,
                key: helper.jobKey(this._setting.queueName, this._setting.prefix, job.id),
                data: job.data,
                type: job.name,
                queueName: job.queue.name,
                prefix: job.queue.keyPrefix,
                done: done
            }
            this.emit('job', res);
        });
    }
}

module.exports = Consumer;
